pkg/core/native_contract_test.go

```diff
@@ -1,324 +1,168 @@
-	if err != nil {
-		panic(err)
-	}
-	bs, err := args[1].TryBytes()
-	if err != nil {
-		panic(err)
-	}
-	err = contract.CallFromNative(ic, tn.meta.Hash, cs, string(bs), args[2].Value().([]stackitem.Item), hasReturn)
-	if err != nil {
-		panic(err)
-	}
-}
-
-func (tn *testNative) callOtherContractNoReturn(ic *interop.Context, args []stackitem.Item) stackitem.Item {
-	tn.call(ic, args, false)
-	return stackitem.Null{}
-}
-
-func (tn *testNative) callOtherContractWithReturn(ic *interop.Context, args []stackitem.Item) stackitem.Item {
-	tn.call(ic, args, true)
-	bi := ic.VM.Estack().Pop().BigInt()
-	return stackitem.Make(bi.Add(bi, big.NewInt(1)))
-}
-
 func TestNativeContract_Invoke(t *testing.T) {
-	chain := newTestChain(t)
-
-	tn := newTestNative()
-	chain.registerNative(tn)
-
-	err := chain.contracts.Management.PutContractState(chain.dao, &state.Contract{
-		ContractBase: state.ContractBase{
-			ID:       1,
-			NEF:      tn.meta.NEF,
-			Hash:     tn.meta.Hash,
-			Manifest: tn.meta.Manifest,
-		},
-	})
-	require.NoError(t, err)
-
-	// System.Contract.Call + "sum" itself + opcodes for pushing arguments.
-	price := int64(testSumCPUFee * chain.GetBaseExecFee() * 2)
-	price += testSumStorageFee * chain.GetStoragePrice()
-	price += 3 * fee.Opcode(chain.GetBaseExecFee(), opcode.PUSHINT8)
-	price += 2 * fee.Opcode(chain.GetBaseExecFee(), opcode.SYSCALL, opcode.PUSHDATA1, opcode.PUSHINT8)
-	price += fee.Opcode(chain.GetBaseExecFee(), opcode.PACK)
-	res, err := invokeContractMethod(chain, price, tn.Metadata().Hash, "sum", int64(14), int64(28))
-	require.NoError(t, err)
-	checkResult(t, res, stackitem.Make(42))
-	_, err = chain.persist(false)
-	require.NoError(t, err)
-
-	select {
-	case index := <-tn.blocks:
-		require.Equal(t, chain.blockHeight, index)
-	default:
-		require.Fail(t, "onPersist wasn't called")
-	}
-
-	// Enough for Call and other opcodes, but not enough for "sum" call.
-	res, err = invokeContractMethod(chain, price-1, tn.Metadata().Hash, "sum", int64(14), int64(28))
-	require.NoError(t, err)
-	checkFAULTState(t, res)
+	const (
+		transferCPUFee          = 1 << 17
+		transferStorageFee      = 50
+		systemContractCallPrice = 1 << 15
+	)
+	bc, validator, committee := chain.NewMulti(t)
+	e := neotest.NewExecutor(t, bc, validator, committee)
+	gasHash := e.NativeHash(t, nativenames.Gas)
+
+	baseExecFee := bc.GetBaseExecFee()
+	price := fee.Opcode(baseExecFee, opcode.SYSCALL, // System.Contract.Call
+		opcode.PUSHDATA1, // contract hash (20 byte)
+		opcode.PUSHDATA1, // method
+		opcode.PUSH15,    // call flags
+		// `transfer` args:
+		opcode.PUSHDATA1, // from
+		opcode.PUSHDATA1, // to
+		opcode.PUSH1,     // amount
+		opcode.PUSHNULL,  // data
+		// end args
+		opcode.PUSH4, // amount of args
+		opcode.PACK,  // pack args
+	)
+	price += systemContractCallPrice*baseExecFee + // System.Contract.Call price
+		transferCPUFee*baseExecFee + // `transfer` itself
+		transferStorageFee*bc.GetStoragePrice() // `transfer` storage price
```

Comment:
Can we just test-invoke for the first invocation and then use GAS value from it?

Reply:
We can, but the aim of this test is to check that amount of GAS required for this call is exactly as we expect it to be, and if we'll use testinvoke as a reference then we won't be sure about that.